### PR TITLE
Support dynamic table

### PR DESCRIPTION
**Pull request by jianyun8023**
Support dynamic table

### CODE REVIEW

File: pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarDynamicTableSource.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        return "Pulsar universal table source";
+    }
+
+    private static ClientConfigurationData newClientConf(String serviceUrl) {
+        ClientConfigurationData clientConf = new ClientConfigurationData();
+        clientConf.setServiceUrl(serviceUrl);
+        return clientConf;
+    }
+
+    private static Properties removeConnectorPrefix(Properties in) {
```
Comment (reswqa) [inline pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarDynamicTableSource.java:184]: This method seems to be repeated to PulsarDynamicTableFactory.class . Can we extract it and put it into the option class or some utils class.

**Review by reswqa — COMMENTED** [review]
LGTM, and If it is convenient, can we add some related test case to the dynamic table.

File: pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarOptions.java [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+					+ "\"fixed\": (each Flink partition ends up in at most one Kafka partition),\n"
+					+ "\"round-robin\": (a Flink partition is distributed to Kafka partitions round-robin)\n"
+					+ "\"custom class name\": (use a custom FlinkKafkaPartitioner subclass)");
+
+	// --------------------------------------------------------------------------------------------
+	// Option enumerations
+	// --------------------------------------------------------------------------------------------
+
+	// Start up offset.
+	public static final String SCAN_STARTUP_MODE_VALUE_EARLIEST = "earliest";
```
[the remark this answers is not in the document]
Comment (jianyun8023) [reply pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarOptions.java:143]: The old source didn't have this feature yet.

File: pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarOptions.java [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+	 * @return SpecificOffsets with Map format, key is partition, and value is offset
+	 */
+	public static Map<Integer, String> parseSpecificOffsets(
+			String specificOffsetsStr,
+			String optionKey) {
+		final Map<Integer, String> offsetMap = new HashMap<>();
+		final String[] pairs = specificOffsetsStr.split(";");
+		final String validationExceptionMessage = String.format(
+				"Invalid properties '%s' should follow the format "
+						+ "'42:1012:0;44:1011:1', but is '%s'.",
```
[the remark this answers is not in the document]
Comment (jianyun8023) [reply pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarOptions.java:346]: I'll add it.

File: pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarOptions.java [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+				optionKey,
+				specificOffsetsStr);
+
+		if (pairs.length == 0) {
+			throw new ValidationException(validationExceptionMessage);
+		}
+
+		for (String pair : pairs) {
+			if (null == pair || pair.length() == 0 || !pair.contains(",")) {
```
[the remark this answers is not in the document]
Comment (jianyun8023) [reply pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarOptions.java:355]: You need to look at the context, two lines of code here doing the validation of illegal parameters.

File: pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarOptions.java [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+				offsetMap.put(partition, offsetValue);
+			} catch (NumberFormatException e) {
+				throw new ValidationException(validationExceptionMessage, e);
+			}
+		}
+		return offsetMap;
+	}
+
+	/** Decides if the table options contains Kafka client properties that start with prefix 'properties'. */
+	private static boolean hasKafkaClientProperties(Map<String, String> tableOptions) {
```
[the remark this answers is not in the document]
Comment (jianyun8023) [reply pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarOptions.java:377]: remove it.

File: pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarDynamicTableSink.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    protected final String serviceUrl;
+    protected final String adminUrl;
+
+    /**
+     * Properties for the Kafka producer.
+     */
+    protected final Properties properties;
+
+    /**
+     * Sink format for encoding records to Kafka.
```
[the remark this answers is not in the document]
Comment (jianyun8023) [reply pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarDynamicTableSink.java:56]: ok

File: pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarDynamicTableSink.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    public ChangelogMode getChangelogMode(ChangelogMode requestedMode) {
+        return this.encodingFormat.getChangelogMode();
+    }
+
+    @Override
+    public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
+        SerializationSchema<RowData> serializationSchema =
+                this.encodingFormat.createRuntimeEncoder(context, this.consumedDataType);
+        final SinkFunction<RowData> pulsarSink = createPulsarSink(
+                this.topic,
```
[the remark this answers is not in the document]
Comment (jianyun8023) [reply pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarDynamicTableSink.java:85]: ok

File: pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarDynamicTableSource.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR;
+import static org.apache.flink.table.descriptors.PulsarValidator.CONNECTOR_EXTERNAL_SUB_DEFAULT_OFFSET;
+import static org.apache.flink.table.descriptors.PulsarValidator.CONNECTOR_STARTUP_MODE_VALUE_EARLIEST;
+
+/**
+ * pulsar dynamic table source.
+ */
+public class PulsarDynamicTableSource implements ScanTableSource {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PulsarDynamicTableSource.class);
```
[the remark this answers is not in the document]
Comment (jianyun8023) [reply pulsar-flink-1.11/src/main/java/org/apache/flink/table/PulsarDynamicTableSource.java:54]: ok

**Review by reswqa — APPROVED** [review]

**Pull request merged**